refactor!: Tidy up AND/OR/NOT API and usage

scovich opened the pull request:
## What changes are proposed in this pull request?

Get rid of `impl Not for Expression` because it's hard to use directly (have to import `std::ops::Not`) and is also forced to take `self` instead of `impl Into<Expression>` as an argument (which we use way more often than `!`). Further, NOT is the only logical operator we could meaningfully overload in rust. AND/OR cannot be overloaded at all, and comparisons (via `PartialEq` and `PartialOrd`) have hard-wired return types that don't work for expressions.

Meanwhile `Expression::and` and `Expression::or` are hard to read as infix operators:
```rust
(a.lt(10).and(b.gt(20)).or(c.eq(42).and(d.le(100)))
```
so make them associated methods taking `impl Into<Expression>` instead of taking `self`:
```rust
Expr::or(
    Expr::and(a.lt(10), b.gt(20)),
    Expr::and(c.eq(42), d.le(100)),
)
```

### This PR affects the following public APIs

Public interfaces and methods.

## How was this change tested?

Renaming refactor. Compilation suffices.

nicklan commented on kernel/src/expressions/mod.rs:
```diff
@@ -540,11 +545,12 @@ pub trait ExpressionTransform<'a> {
     }
 }
 
+#[cfg(test)]
```
should we have a `TODO` to remove this once we get it out of tests?

zachschuermann reviewed (approved):
LGTM

zachschuermann commented:
> @zachschuermann -- yet another false negative semver check?? I removed a public trait impl for a public enum, and also changed the signature of a public method.

wow yea that's... concerning

scovich commented:
> > @zachschuermann -- yet another false negative semver check?? I removed a public trait impl for a public enum, and also changed the signature of a public method.
> 
> wow yea that's... concerning

Is there a way to file a bug against semver? This is a public PR, they can easily check out the exact version of the code that caused the problem as a repro.
